src/extension.ts

```diff
@@ -329,35 +341,53 @@ async function runJavaProject(node: any, noDebug: boolean) {
         throw error;
     }
 
-    const mainClassesOptions: IMainClassOption[] = await utility.searchMainMethods(vscode.Uri.parse(node.uri));
-    if (!mainClassesOptions || !mainClassesOptions.length) {
-        vscode.window.showErrorMessage(`Failed to ${noDebug ? "run" : "debug"} this project '${node._nodeData.displayName || node.name}' `
-            + "because it does not contain any main class.");
-        return;
-    }
+    const progressReporter = progressReporterManager.create(noDebug ? "Run" : "Debug");
+    try {
+        progressReporter.report("Resolve mainClass", "Resolving main class...");
+        const mainClassesOptions: IMainClassOption[] = await utility.searchMainMethods(vscode.Uri.parse(node.uri));
+        if (progressReporter.isCancelled()) {
+            throw new utility.OperationCancelledError("");
+        }
 
-    const pick = await mainClassPicker.showQuickPickWithRecentlyUsed(mainClassesOptions,
-        "Select the main class to run.");
-    if (!pick) {
-        return;
-    }
+        if (!mainClassesOptions || !mainClassesOptions.length) {
+            vscode.window.showErrorMessage(`Failed to ${noDebug ? "run" : "debug"} this project '${node._nodeData.displayName || node.name}' `
+                + "because it does not contain any main class.");
+            throw new utility.OperationCancelledError("");
+        }
 
-    const projectName: string | undefined = pick.projectName;
-    const mainClass: string = pick.mainClass;
-    const filePath: string | undefined = pick.filePath;
-    const workspaceFolder: vscode.WorkspaceFolder | undefined = filePath ? vscode.workspace.getWorkspaceFolder(vscode.Uri.file(filePath)) : undefined;
-    const launchConfigurations: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration("launch", workspaceFolder);
-    const existingConfigs: vscode.DebugConfiguration[] = launchConfigurations.configurations;
-    const existConfig: vscode.DebugConfiguration | undefined = _.find(existingConfigs, (config) => {
-        return config.mainClass === mainClass && _.toString(config.projectName) === _.toString(projectName);
-    });
-    const debugConfig = existConfig || {
-        type: "java",
-        name: `Launch - ${mainClass.substr(mainClass.lastIndexOf(".") + 1)}`,
-        request: "launch",
-        mainClass,
-        projectName,
-    };
-    debugConfig.noDebug = noDebug;
-    vscode.debug.startDebugging(workspaceFolder, debugConfig);
+        progressReporter.report("Select mainClass", "Selecting the main class to run...");
```

Comment:
Here is my first thinking when I saw the message here.

Basically the messages are used to telling the user what the `extension` is doing right now. So it seems that it's telling the user the extension is selecting the main class, but actually it's waiting for the user to select.